transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,471 @@
+                match event {
+                    Some(ToEndpoint::Dial {
+                        addr,
+                        public_key,
+                        tx,
+                    }) => {
+                        let crypto = C::new_client_config(&me.crypto_config, public_key);
+                        let client_config = ClientConfig {
+                            transport: me.crypto_config.transport.clone(),
+                            crypto,
+                        };
+                        let (id, connection) =
+                            match me.endpoint.connect(client_config, addr, "server_name") {
+                                Ok(c) => c,
+                                Err(err) => {
+                                    tracing::error!("dial failure: {}", err);
+                                    let _ = tx.send(Err(err.into()));
+                                    continue;
+                                }
+                            };
+                        let (channel, conn) = me.channel.create_connection(id);
+                        me.connections.insert(id, conn);
+                        let muxer = QuicMuxer::new(channel, connection);
+                        tx.send(Ok(muxer)).ok();
+                    }
+                    Some(ToEndpoint::ConnectionEvent {
+                        connection_id,
+                        event,
+                    }) => {
+                        let is_drained_event = event.is_drained();
+                        if is_drained_event {
+                            me.connections.remove(&connection_id);
+                        }
+                        if let Some(event) = me.endpoint.handle_event(connection_id, event) {
+                            if me.send_event(connection_id, event, cx) {
+                                tracing::info!("filled event slot");
+                                break;
+                            }
+                        }
+                    }
+                    Some(ToEndpoint::Transmit(transmit)) => {
+                        me.transmit(transmit);
+                    }
+                    None => {
+                        me.endpoint.reject_new_connections();
+                        return Poll::Ready(());
+                    }
+                }
+            }
+        }
+
+        while !me.outgoing.is_empty() {
+            let transmits: &[_] = me.outgoing.make_contiguous();
+            match me.socket.poll_send(cx, transmits) {
+                Poll::Ready(Ok(n)) => {
+                    me.outgoing.drain(..n);
+                }
+                Poll::Ready(Err(err)) => tracing::error!("send_to: {}", err),
+                Poll::Pending => break,
+            }
+        }
+
+        if me.event_slot.is_none() && me.incoming_slot.is_none() {
+            let mut metas = [RecvMeta::default(); BATCH_SIZE];
+            let mut iovs = MaybeUninit::<[IoSliceMut; BATCH_SIZE]>::uninit();
```

Comment:
I am assuming this is an optimization. If so, given the usage of unsafe code, what performance impact did this optimization show in benchmarks?